common/http.hpp

```diff
@@ -236,90 +237,62 @@ class OutgoingRequest : public Request {
 
 class Stream;
 
-class IncomingRequest : public Request {
+class IncomingRequest :
+	public Request,
+	virtual public io::Canceller,
+	public enable_shared_from_this<IncomingRequest> {
 public:
+	~IncomingRequest();
+
 	// Set this after receiving the headers, if appropriate.
 	void SetBodyWriter(io::WriterPtr body_writer);
-	io::AsyncReaderPtr MakeBodyAsyncReader();
+	io::ExpectedAsyncReaderPtr MakeBodyAsyncReader();
 
 	// Use this to get a response that can be used to reply to the request. Due to the
 	// asynchronous nature, this can be done immediately or some time later.
 	ExpectedOutgoingResponsePtr MakeResponse();
 
-	void Cancel();
+	void Cancel() override;
 
 private:
 	IncomingRequest(weak_ptr<Stream> stream) :
 		stream_(stream) {
 	}
 
-	class BodyAsyncReader : virtual public io::AsyncReader {
-	public:
-		BodyAsyncReader(weak_ptr<Stream> stream);
-		~BodyAsyncReader();
-
-		error::Error AsyncRead(
-			vector<uint8_t>::iterator start,
-			vector<uint8_t>::iterator end,
-			io::AsyncIoHandler handler) override;
-		void Cancel() override;
-
-	private:
-		weak_ptr<Stream> stream_;
-		bool done_ {false};
-
-		friend class Stream;
-	};
-
 	weak_ptr<Stream> stream_;
 
-	io::WriterPtr body_writer_;
-	shared_ptr<BodyAsyncReader> body_async_reader_;
-
+	friend class Server;
 	friend class Stream;
 };
 
-class IncomingResponse : public Response {
+class IncomingResponse :
+	public Response,
+	virtual public io::Canceller,
+	public enable_shared_from_this<IncomingResponse> {
 public:
+	void Cancel() override;
+
 	// Use these after receiving the headers, if appropriate.
 	void SetBodyWriter(io::WriterPtr body_writer);
-	io::AsyncReaderPtr MakeBodyAsyncReader();
+	io::ExpectedAsyncReaderPtr MakeBodyAsyncReader();
```

Comment:
Rephrase the `if appropriate` part  to reflect that these are not optional anymore (the user needs to use one of the two methods below).

Reply:
This applies to`IncomingRequest` and other places too.